Add ephemeral storage e2e tests

NickrenREN opened the pull request:
Add e2e tests of limitrange/quota/downward_api for local ephemeral storage

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: part of #52463

**Special notes for your reviewer**:
Add e2e tests of limitrange/quota/downwardapi for local ephemeral storage

**Release note**:
```release-note
Add limitrange/resourcequota/downward_api  e2e tests for local ephemeral storage
```

/assign @jingxu97

Commit:
Add limitrange e2e test for LocalStorageCapacityIsolation feature

Commit:
Add resource quota e2e test for LocalStorageCapacityIsolation feature

Commit:
Add downward_api e2e test for LocalStorageCapacityIsolation feature

NickrenREN commented:
/retest

NickrenREN commented:
/retest

jingxu97 commented on test/e2e/common/downward_api.go:
```diff
@@ -183,6 +183,82 @@ var _ = framework.KubeDescribe("Downward API", func() {
+				},
+			}
+			expectations := []string{
+				fmt.Sprintf("EPHEMERAL_STORAGE_LIMIT=67108864"),
```
so 67108864 is calculated from 64Mi? Can we change to calculate this number instead of hardcode it here?

NickrenREN replied:
You mean change 67108864 to 64*1024*1024 right?

NickrenREN replied:
Done, @jingxu97

NickrenREN commented:
/retest

NickrenREN commented:
/retest

jingxu97 commented:
cc @jeffvance

jeffvance commented on test/e2e/common/downward_api.go:
```diff
@@ -183,6 +183,82 @@ var _ = framework.KubeDescribe("Downward API", func() {
+			framework.SkipUnlessLocalEphemeralStorageEnabled()
+		})
+
+		It("should provide container's limits.ephemeral-storage and requests.ephemeral-storage as env vars [Feature:LocalStorageCapacityIsolation]", func() {
```
It seems that `[Feature:.*]` here is redundant with `Context` description above.

NickrenREN replied:
yeah, i added `[Feature:.*]` first but it did not work, so i added `Context` later

jeffvance replied:
The `Context` text is not written to the output at the same time as `Describe` and `It` text, but you will see it in the test result, and the context text is included in the `focus=` filter. I move common `It` text to the parent `Context` description just so it isn't repeated. But what you have here works fine too.

jeffvance replied:
My above comment has not been addressed.

jeffvance commented on test/e2e/common/downward_api.go:
```diff
@@ -183,6 +183,82 @@ var _ = framework.KubeDescribe("Downward API", func() {
+			testDownwardAPIForEphemeralStorage(f, podName, env, expectations)
+		})
+
+		It("should provide default limits.ephemeral-storage from node allocatable [Feature:LocalStorageCapacityIsolation]", func() {
```
ditto on [Feature...]

jeffvance commented on test/e2e/common/downward_api.go:
```diff
@@ -183,6 +183,82 @@ var _ = framework.KubeDescribe("Downward API", func() {
+				},
+			}
+			expectations := []string{
+				fmt.Sprintf("EPHEMERAL_STORAGE_LIMIT=64*1024*1024"),
```
why use `fmt.Sprintf` here and below?

NickrenREN replied:
sorry, copy from the exsiting code directly

NickrenREN replied:
Modified

jeffvance commented on test/e2e/common/downward_api.go:
```diff
@@ -217,6 +293,36 @@ func testDownwardAPI(f *framework.Framework, podName string, env []v1.EnvVar, ex
+	testDownwardAPIUsingPod(f, pod, env, expectations)
+}
+
 func testDownwardAPIUsingPod(f *framework.Framework, pod *v1.Pod, env []v1.EnvVar, expectations []string) {
```
where is `env` used here?

NickrenREN replied:
yeah, seems not used, it is not added by me, i will delete it

jeffvance replied:
I still see `env`...

jeffvance replied:
This comment appears to not have been addressed.

NickrenREN replied:
Yeah, this is not introduced by me. I will look into that and send a separate pr to solve it

jeffvance commented on test/e2e/limit_range.go:
```diff
@@ -93,17 +93,122 @@ var _ = framework.KubeDescribe("LimitRange", func() {
+		framework.SkipUnlessLocalEphemeralStorageEnabled()
+	})
+
+	It("should create a LimitRange with default ephemeral storage and ensure pod has the default applied.", func() {
```
A large amount of the code below is duplicated from the It block above. It seems reasonable to either use a

NickrenREN replied:
For now, since the ephemeral storage feature is alpha, i add a new  `KubeDescribe` block, and when the feature graduates to beta or GA, i will merge them, should we do that now ?

jeffvance replied:
I guess it is not needed now unless you think there is a chance that after GA you won't have time to consolidate the code. It is not critical but it usually preferred to extract out common patterns (DRY).

jeffvance commented on test/e2e/limit_range.go:
```diff
@@ -93,17 +93,122 @@ var _ = framework.KubeDescribe("LimitRange", func() {
+		Expect(err).To(HaveOccurred())
+	})
+
+})
```
where are the pod created above deleted?

NickrenREN replied:
In framework `AfterEach` function

jeffvance replied:
In the `AfterEach` where the namespace is deleted? If yes, it is far more common to _explicitly_ perform all cleanup in the test itself, even though the namespaced resources are deleted when the namespace is deleted.

NickrenREN replied:
Done

jeffvance replied:
ack

jeffvance commented on test/e2e/resource_quota.go:
```diff
@@ -41,6 +41,75 @@ const (
 
 var classGold string = "gold"
 
+var _ = framework.KubeDescribe("ResourceQuota [Feature:LocalStorageCapacityIsolation]", func() {
```
why is this new code a separate `KubeDescribe`? I'm not sure it matters much, but rather than a new KubeDescribe I think you could have added `Context()`s under the existing KubeDesribe.

NickrenREN replied:
yeah, same reason to limitrange, i want to merge them together when the feature graduates to beta or GA

jeffvance commented on test/e2e/resource_quota.go:
```diff
@@ -41,6 +41,75 @@ const (
+		framework.SkipUnlessLocalEphemeralStorageEnabled()
+	})
+
+	It("should create a ResourceQuota and capture the life of a pod.", func() {
```
Other than ~6 lines this `It` block is a duplicate of the same test case below. Is there a way to consolidate these 2 It blocks? Maybe a table-driven tests or helper funcs?

jeffvance commented on test/e2e/resource_quota.go:
```diff
@@ -41,6 +41,75 @@ const (
+		usedResources[v1.ResourceEphemeralStorage] = resource.MustParse("0")
+		err = waitForResourceQuota(f.ClientSet, f.Namespace.Name, quotaName, usedResources)
+		Expect(err).NotTo(HaveOccurred())
+	})
```
does the ResourceQuota need to be deleted?

jeffvance replied:
Though the ResourceQuote will be deleted when the framework's AfterEach is invoked, it is more typical in the storage e2e tests to explicitly delete all resources that the test creates.

NickrenREN replied:
Done

jeffvance commented on test/e2e/resource_quota.go:
```diff
@@ -41,6 +41,75 @@ const (
+		requests[v1.ResourceEphemeralStorage] = resource.MustParse("100Mi")
+		podToUpdate.Spec.Containers[0].Resources.Requests = requests
+		_, err = f.ClientSet.CoreV1().Pods(f.Namespace.Name).Update(podToUpdate)
+		Expect(err).To(HaveOccurred())
```
note: pod value is now stale

jeffvance commented on test/e2e/resource_quota.go:
```diff
@@ -41,6 +41,75 @@ const (
+		requests = v1.ResourceList{}
+		requests[v1.ResourceEphemeralStorage] = resource.MustParse("100Mi")
+		podToUpdate.Spec.Containers[0].Resources.Requests = requests
+		_, err = f.ClientSet.CoreV1().Pods(f.Namespace.Name).Update(podToUpdate)
```
note: pod value is now stale

jeffvance commented on test/e2e/resource_quota.go:
```diff
@@ -41,6 +41,75 @@ const (
+		By("Not allowing a pod to be created that exceeds remaining quota")
+		requests = v1.ResourceList{}
+		requests[v1.ResourceEphemeralStorage] = resource.MustParse("300Mi")
+		pod = newTestPodForQuota(f, "fail-pod", requests, v1.ResourceList{})
```
note: pod value is now stale

NickrenREN replied:
Sorry, i do not understand, this is a new pod.

jeffvance replied:
Bad comment. Never mind.

jeffvance commented:
Rebase needed.

NickrenREN commented:
@jeffvance thanks for your review. I have deleted the resources(`pod/limitrange/resourcequota`) explicitly it their test and hope i have addressed your comments. PTAL

jeffvance commented on test/e2e/resource_quota.go:
```diff
@@ -41,6 +41,79 @@ const (
+		deleteResourceQuota(f.ClientSet, f.Namespace.Name, resourceQuota.Name)
+		Expect(err).NotTo(HaveOccurred())
+	})
+})
```
About deleting resources created by the test suite: since the test could fail at any point after a resource is created, and we want the test to cleanup its environment explicitly (rather than relying on the generated namespace being deleted), a `defer` after the resource is created or an `AfterEach` in the `Context` (or `Describe`) are commonly used for this purpose.

NickrenREN replied:
I found we did the deletion in the test in other places like resourcequota(cpu,mem part).
If we want to delete all the pods and other resources in AfterEach function, should we use different var names to represent the pods created in the test So that we can delete all of them?

jeffvance replied:
If I understand your question, you need the pod's name to Delete it. This can be done in an `AfterEach` but then you'll need the scope of the `podName` variable to be global to the parent `Context` or `Describe`. If you delete the pod in a `defer` then you can create the pod value in the `It` block and defer-delete it after the pod-Create is successful.
If I understand your question, you can delete a pod in an `AfterEach` but you'll need the scope of the `podName` variable to be global to the parent `Context` or `Describe`. If you delete the pod in a `defer` then you create the pod value in the `It` block and defer-delete it after the pod-Create is successful.

NickrenREN replied:
done

NickrenREN commented:
@jeffvance @jingxu97  Addressed your comments and rebased. PTAL, thanks

NickrenREN commented:
/retest

NickrenREN commented:
/retest

jeffvance commented on test/e2e/common/downward_api.go:
```diff
@@ -225,6 +225,81 @@ var _ = Describe("[sig-api-machinery] Downward API", func() {
+				},
+			}
+			expectations := []string{
+				fmt.Sprintf("EPHEMERAL_STORAGE_LIMIT=[1-9]"),
```
why use fmt.Sprintf here?

NickrenREN replied:
yeah, no need. removed

jeffvance commented on test/e2e/scheduling/resource_quota.go:
```diff
@@ -41,6 +41,83 @@ const (
+
+		defer func() {
+			By("Removing resourceQuota")
+			deleteResourceQuota(f.ClientSet, f.Namespace.Name, resourceQuota.Name)
```
func return not assigned to `err`

NickrenREN replied:
Good catch. thanks, updated

NickrenREN commented:
/retest

NickrenREN commented:
/retest

jeffvance commented on test/e2e/scheduling/limit_range.go:
```diff
@@ -108,6 +108,137 @@ var _ = SIGDescribe("LimitRange", func() {
+		Expect(err).NotTo(HaveOccurred())
+
+		defer func() {
+			By("Removing pod")
```
maybe add pod.Name in the `By` description?

jeffvance commented on test/e2e/scheduling/limit_range.go:
```diff
@@ -108,6 +108,137 @@ var _ = SIGDescribe("LimitRange", func() {
+
+		defer func() {
+			By("Removing pod")
+			err = f.ClientSet.CoreV1().Pods(f.Namespace.Name).Delete(pod.Name, nil)
```
are you sure this pod is deleted vs. the pod created below and assigned to the same pointer variable?

jeffvance commented on test/e2e/scheduling/resource_quota.go:
```diff
@@ -45,6 +45,83 @@ var classGold string = "gold"
+	})
+})
+
+var _ = SIGDescribe("ResourceQuota", func() {
```
why is the same SIGDescribe used here too?

NickrenREN replied:
This is for alpha feature and will skip if local ephemeral storage feature is not enabled in BeforeEach.
As i said before, will merge it when the feature graduates to beta or GA

jeffvance commented:
/lgtm

NickrenREN commented:
/assign @thockin @smarterclayton

NickrenREN commented:
/retest

NickrenREN commented:
/retest

NickrenREN commented:
/test all

NickrenREN commented:
/retest

NickrenREN commented:
It seems that `retest` command does not work

NickrenREN commented:
/retest all

NickrenREN commented:
/retest

thockin commented:
/approve
